hack/make-rules/test-cmd-util.sh

```diff
@@ -1621,6 +1645,31 @@ run_non_native_resource_tests() {
   # Test that we can list this new CustomResource (bars)
   kube::test::get_object_assert bars "{{range.items}}{{$id_field}}:{{end}}" ''
 
+  # Test that we can list this new CustomResource (resources)
+  kube::test::get_object_assert resources "{{range.items}}{{$id_field}}:{{end}}" '////--'
+
+  # Test that we can create a new resource of type Kind
+  kubectl "${kube_flags[@]}" create -f hack/testdata/CRD/resource.yaml "${kube_flags[@]}"
+
+  # Test that we can list this new custom resource
+  kube::test::get_object_assert foos "{{range.items}}{{$id_field}}:{{end}}" 'test:'
+
+  # Test that -o name returns kind.group/resourcename
+  output_message=$(kubectl "${kube_flags[@]}" get resource/myobj -o name)
+  kube::test::if_has_string "${output_message}" 'kind.mygroup.example.com/myobj'
+
+  output_message=$(kubectl "${kube_flags[@]}" get resources/myobj -o name)
```

Comment:
@deads2k I omitted a test for `$ kubectl get kind/myobj` as it currently fails with:
```
$ kubectl get kind/myobj -o name
the server doesn't have a resource type "kind"
```

When only a kind is given, we hit [this condition](https://github.com/kubernetes/kubernetes/pull/59353/files#diff-80a5c1e5d1429094c6ecb3ae22847912R615) in the resource builder, where we ask the mapper for a restmapping, using a GroupKind with no group, and an empty version.

We end up going through the [priority restmapper's RESTMapping method](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/priority.go#L156), which in turn calls the [default restmapper's RESTMappings method](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/restmapper.go#L485). 

That method attempts to iterate through any versions given (none in this case), and then attempts to use the default groupversions which were [passed to the instance here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/restmapper.go#L92) in order to find any group versions with a matching group name.

Since we have an empty group in this case (only using kind), this method then [proceeds to fail](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/restmapper.go#L513), causing us to ultimately end up with the error from the example above.

[I have a patch here](https://gist.github.com/juanvallejo/1c77c14d18b14f15c09377cff0a02369) that iterates through stored gvks on the restmapper instance (added through the [AddSpecific method here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/restmapper.go#L110)) and does the same thing that we already attempt to do with groups: iterates through stored gvks, looking for ones that have matching `kind`s.

I have tested this patch locally with [Jordan's example](https://github.com/kubernetes/kubernetes/pull/59227#issuecomment-363173986), and works as expected when attempting to fetch resources by only specifying their kind. PTAL

Reply:
Isn't failure there exactly what we expect on the kind match (which shouldn't match because that isn't the kind group) and the code should fail through to the resource matching?  Why isn't it doing that?

Reply:
> @deads2k I omitted a test for $ kubectl get kind/myobj as it currently fails with:

That means this pull is broken.  You should be up front about the pull be broken by putting the broken test in your pull.

Reply:
> Isn't failure there exactly what we expect on the kind match (which shouldn't match because that isn't the kind group) and the code should fail through to the resource matching? Why isn't it doing that?

It is doing that. The code for resource is what ultimately ends up returning the error message `the server doesn't have a resource type "kind"`. I have added a test for `kubectl get kind -o name`

Reply:
> @deads2k I omitted a test for $ kubectl get kind/myobj as it currently fails with:
> 
> $ kubectl get kind/myobj -o name
> the server doesn't have a resource type "kind"

kind is always paired with group for all non-legacy kinds.  Legacy kinds all match their resources.  We don't have to worry about this case.